<div style="border:solid green 2px; padding: 20px"> 
    
Hello Harvey, 
    
    
my name is Aleksandra. I'm going to review your project! Nice to meet you! 🙌

You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments
</div>

Please don't remove my comments :) If you have any questions don't hesitate to respond to my comments in a different section. 

<div class="alert alert-block alert-success">
<b>Overall reviewer's comment v1</b> <a class="tocSkip"></a>
    
Harvey, thanks a lot for the project! Great job! Everything is perfect. I can see that a lot of efforts have been made, and it turned out to be very strong. You work very carefully with data and you make qualitative assumptions - like a real analyst. There are no critical comments. I left guidance ones as we went along. If you have any questions,  feel free to ask me through your tutor. Good luck at next steps!

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.

[In this notebook you're provided with hints and brief instructions and thinking prompts. Don't ignore them as they are designed to equip you with the structure for the project and will help you analyze what you're doing on a deeper level. Before submitting your project, make sure you remove all hints and descriptions provided to you. Instead, make this report look as if you're sending it to your teammates to demonstrate your findings - they shouldn't know you had some external help from us! To help you out, we've placed the hints you should remove in square brackets.]

[Before you dive into analyzing your data, explain the purposes of the project and hypotheses you're going to test.]

## Open the data file and have a look at the general information. 
Import Panda libraries

In [1]:
# Loading all the libraries
import pandas as pd

# Load the data
df = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

Print out the dataset to have an overview and understand it better

In [2]:
# Let's see how many rows and columns our dataset has

print(df)

       children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status

In [3]:
# let's print the first 10 rows

print(df.head(10))

   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1

The days_employed column seems to have negative values. Other than that, the dataset seems ok at first glance. But it'd serve us well to look at the dataset info.

In [4]:
# Get info on data
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


Just as we suspected. There are missing values in the dataset. We can see them in total_income and days_employed columns.

In [5]:
# Let's look at the filtered table with missing values in the the first column with missing data

print(df[df['days_employed'].isna()])

       children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status

It seems the rows with missing data are identical in both columns. But we have to check if the number of rows of missing data are the same in both columns to be able to draw a conclusion.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b>  You are right

In [6]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.

days_employed_filtered = (df[df['days_employed'].isna()])
total_income_filtered = (df[df['total_income'].isna()])
days_employed_percentage = len(days_employed_filtered)/ len(df['days_employed'])
total_income_percentage = len(total_income_filtered)/len(df['total_income'])
print(len(days_employed_filtered))
print(len(total_income_filtered))

print(f'Percentage of missing values in days employed column: {days_employed_percentage:.1%}')
print(f'Percentage of missing values in total income column: {total_income_percentage:.1%}')

2174
2174
Percentage of missing values in days employed column: 10.1%
Percentage of missing values in total income column: 10.1%


**Intermediate conclusion**

The rows with mising data are of the same number. Which means that for every row where days employed was missed, total income was also missed. Could it mean that these customers are unemployed?

The percentage of missing data in each column is 10.1%. Though that may not be a lot, it's worth looking into. 

Let's check if there's truly a relationship between income type and the missing values.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> Yes, the amount of missing values is significant. We can't delete them

In [7]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
print(days_employed_filtered.groupby('income_type').count())



               children  days_employed  dob_years  education  education_id  \
income_type                                                                  
business            508              0        508        508           508   
civil servant       147              0        147        147           147   
employee           1105              0       1105       1105          1105   
entrepreneur          1              0          1          1             1   
retiree             413              0        413        413           413   

               family_status  family_status_id  gender  debt  total_income  \
income_type                                                                  
business                 508               508     508   508             0   
civil servant            147               147     147   147             0   
employee                1105              1105    1105  1105             0   
entrepreneur               1                 1       1     1   

In [8]:
# Checking distribution

print(days_employed_filtered.pivot_table(index='income_type', values='days_employed', aggfunc='count'))

               days_employed
income_type                 
business                   0
civil servant              0
employee                   0
entrepreneur               0
retiree                    0


From what we could gather, there doesn't seem to be a relationship between specific income types and the missing values.

**Possible reasons for missing values in data**

We still stand by the assumption that customers with missing values may be unemployed. Although, there doesn't seem to be any patterns to the missing data.

In [9]:
# Checking the distribution in the whole dataset

print(df.pivot_table(index='income_type', values='days_employed', aggfunc='count'))

                             days_employed
income_type                               
business                              4577
civil servant                         1312
employee                             10014
entrepreneur                             1
paternity / maternity leave              1
retiree                               3443
student                                  1
unemployed                               2


**Intermediate conclusion**
From our assessment, the missing data seems to be random

In [10]:
# Check for other reasons and patterns that could lead to missing values



**Intermediate conclusion**

None

In [11]:
# Checking for other patterns - explain which

**Conclusions**

There was no identifiable pattern to the missing data

We might need to fill missing values with the mean or median of their columns to be able to draw good conclusions on the dataset.

Next we will explore each column and make corrections to the data if need be.

## Data transformation

Let's go through each column to see what issues we may have in them.

Begin with removing duplicates and fixing educational information if required.

In [12]:
# Let's see all values in education column to check if and what spellings will need to be fixed
print(df['education'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


In [13]:
# Fix the registers if required
df['education']= df['education'].str.upper()

<div class="alert alert-block alert-warning">
<b>Reviewer's comment v1</b> It's always better to lowercase data

In [14]:
# Checking all the values in the column to make sure we fixed them

print(df['education'].unique())

["BACHELOR'S DEGREE" 'SECONDARY EDUCATION' 'SOME COLLEGE'
 'PRIMARY EDUCATION' 'GRADUATE DEGREE']


Check the data the `children` column

In [15]:
# Let's see the distribution of values in the `children` column
print(df['children'].unique())
print(len(df[df['children'] == -1])/len(df['children']))

[ 1  0  3  2 -1  4 20  5]
0.002183507549361208


From the `children` column, we can see a rather interesting value of -1. This may be a typo or may mean the child is deceased but there's no way to know for sure. However, It's percentage is negligible so it won't make much of a difference. I suggest we look at it as a typo and correct it. Theres also a value with 20. This is possible in some parts of the world so it might be safer to leave it.

In [16]:
# [fix the data based on your decision]
df[df['children'] == -1] = 1

<div class="alert alert-warning">
<b>Reviewer's comment v1</b> Great that you calculate share of abnormal values! Now the proportion of anomalies is small and it will not affect anything. Next time, try to calculate their share immediately and if it is less than 5%, get rid of it. Maybe our assumption that  -1==1 is wrong

In [17]:
# Checking the `children` column again to make sure it's all fixed

print(df['children'].unique())

[ 1  0  3  2  4 20  5]


Check the data in the `days_employed` column.

In [18]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
values_percentage = len(df[df['days_employed'] < 0])/len(df['days_employed'])
print(f'{values_percentage:.1%}')

73.7%


A staggering 74.1% of values are negative values and this might have happened due to an error in the system. Let's convert these values to positive values.

In [19]:
# Address the problematic values, if they exist
df['days_employed'] = abs(df['days_employed'])


In [20]:
# Check the result - make sure it's fixed
print(df[df['days_employed'] < 0])
print(df['days_employed'])

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []
0          8437.673028
1          4024.803754
2          5623.422610
3          4124.747207
4        340266.072047
             ...      
21520      4529.316663
21521    343937.404131
21522      2113.346888
21523      3112.481705
21524      1984.507589
Name: days_employed, Length: 21525, dtype: float64


Problem solved.

In [21]:
# Check the `dob_years` for suspicious values and count the percentage

print(df['dob_years'].unique())
print(len(df[df['dob_years'] < 2])/len(df['dob_years']))

[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66  1 69 19 72 70 74 75]
0.006875725900116144


We can see some ages don't match up. Some customers are 0 or 1 year olds and this is quite unlikely. But seeing as the percentage of such an incident is low, it;s best to leave it.

In [22]:
# Address the issues in the `dob_years` column, if they exist
print(df['dob_years'][df['dob_years'] < 2])

99       0
149      0
270      0
291      1
578      0
        ..
20577    0
20717    1
21140    1
21179    0
21313    0
Name: dob_years, Length: 148, dtype: int64


In [23]:
# Check the result - make sure it's fixed


Now let's check the `family_status` column.

In [24]:
# Let's see the values for the column

print(df['family_status'].unique())

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried' 1]


In [25]:
# Address the problematic values in `family_status`, if they exist
print(df.pivot_table(index='family_status', values='family_status_id'))
df['family_status'][df['family_status'] == 1] = 'civil partnership'

                   family_status_id
family_status                      
1                                 1
civil partnership                 1
divorced                          3
married                           0
unmarried                         4
widow / widower                   2


/tmp/ipykernel_54/25736377.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['family_status'][df['family_status'] == 1] = 'civil partnership'


In [26]:
# Check the result - make sure it's fixed
print(df['family_status'].unique())

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']


Now let's check the `gender` column.

In [27]:
# Let's see the values in the column
print(df['gender'].unique())

['F' 'M' 1 'XNA']


In [28]:
# Address the problematic values, if they exist
df['gender'][df['gender'] == 1] = 'N/A'
df['gender'][df['gender'] == 'XNA'] = 'N/A'

/tmp/ipykernel_54/969072300.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'][df['gender'] == 1] = 'N/A'
/tmp/ipykernel_54/969072300.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['gender'][df['gender'] == 'XNA'] = 'N/A'


In [29]:
# Check the result - make sure it's fixed

print(df['gender'].unique())

['F' 'M' 'N/A']


<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> Great that you found that abnormal values!

Now let's check the `income_type` column.

In [30]:
# Let's see the values in the column
print(df['income_type'].unique())

['employee' 'retiree' 'business' 'civil servant' 1 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']


In [31]:
# Address the problematic values, if they exist
df['income_type'][df['income_type'] == 1] = 'unknown'

/tmp/ipykernel_54/2097227964.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['income_type'][df['income_type'] == 1] = 'unknown'


In [32]:
# Check the result - make sure it's fixed

print(df['income_type'].unique())

['employee' 'retiree' 'business' 'civil servant' 'unknown' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']


Now let's see if we have any duplicates in our data.

In [33]:
# Checking duplicates

print(df[df.duplicated() == True])

       children  days_employed  dob_years            education  education_id  \
705           1            1.0          1                    1             1   
742           1            1.0          1                    1             1   
800           1            1.0          1                    1             1   
941           1            1.0          1                    1             1   
1363          1            1.0          1                    1             1   
...         ...            ...        ...                  ...           ...   
21032         0            NaN         60  SECONDARY EDUCATION             1   
21132         0            NaN         47  SECONDARY EDUCATION             1   
21140         1            1.0          1                    1             1   
21281         1            NaN         30    BACHELOR'S DEGREE             0   
21415         0            NaN         54  SECONDARY EDUCATION             1   

           family_status  family_status

In [34]:
# Address the duplicates, if they exist
df=df.drop_duplicates().reset_index(drop=True)

In [35]:
# Last check whether we have any duplicates
print(df[df.duplicated() == True])

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


In [36]:
# Check the size of the dataset that you now have after your first manipulations with it
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21408 entries, 0 to 21407
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21408 non-null  int64  
 1   days_employed     19308 non-null  float64
 2   dob_years         21408 non-null  int64  
 3   education         21408 non-null  object 
 4   education_id      21408 non-null  int64  
 5   family_status     21408 non-null  object 
 6   family_status_id  21408 non-null  int64  
 7   gender            21408 non-null  object 
 8   income_type       21408 non-null  object 
 9   debt              21408 non-null  int64  
 10  total_income      19308 non-null  float64
 11  purpose           21408 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


The duplicates have been dropped. 117 rows which is not up to 10% of our data. The integrity of the data is still preserved and we can move on with our analysis.


<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> You can only delete obvious duplicates after full data processing. Right!

# Working with missing values

To speed up working with some data, some dictionaries could be created for columns containing Id numbers to make it easier to get certain data. But seeing as our dataset is not too big, we may not need them here.

In [37]:
# Find the dictionaries

<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> You are right

### Restoring missing values in `total_income`

As previously stated `days_employed` and `total_income` columns both have missing values and we can fill this using the mean or median of these columns.


Let's start with addressing `total_income` missing values. Let's create an age category for clients. Create a new column with the age category. 
We can split them into active workers and pensioners seeing that the official age of retirement is 65.


In [38]:
# Let's write a function that calculates the age category

def age_category(age):
        if age <= 65:
            return 'Active'
        if age >= 65:
            return 'Pensioner'

In [39]:
# Test if the function works
print(age_category(75))

Pensioner


In [40]:
# Creating new column based on function

df['client_age_group'] = df['dob_years'].apply(age_category)

In [41]:
# Checking how values in the new column
print(df['client_age_group'])


0           Active
1           Active
2           Active
3           Active
4           Active
           ...    
21403       Active
21404    Pensioner
21405       Active
21406       Active
21407       Active
Name: client_age_group, Length: 21408, dtype: object


Let's find the mean and median for the two age groups

Let's create a table that only has data without missing values. This data will be used to restore the missing values.

In [42]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
df_filtered = df[df.isna() == False]
print(df_filtered.head())

   children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    BACHELOR'S DEGREE             0   
1         1    4024.803754         36  SECONDARY EDUCATION             1   
2         0    5623.422610         33  SECONDARY EDUCATION             1   
3         3    4124.747207         32  SECONDARY EDUCATION             1   
4         0  340266.072047         53  SECONDARY EDUCATION             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose client_age_group  
0    purch

In [43]:
# Look at the mean values for income based on your identified factors
pensioner_category = df_filtered[df_filtered['client_age_group'] == 'Pensioner']
pensioner_mean = pensioner_category['total_income'].mean()
print(f'Pensioner mean: {pensioner_mean}')
active_category = df_filtered[df_filtered['client_age_group'] == 'Active']
active_mean = active_category['total_income'].mean()
print(f'Active mean: {active_mean}')

Pensioner mean: 20807.82241679627
Active mean: 26997.246699062416


In [44]:
# Look at the median values for income based on your identified factors
pensioner_median = pensioner_category['total_income'].median()
print(f'Pensioner median: {pensioner_median}')
active_median = active_category['total_income'].median()
print(f'Active median: {active_median}')

Pensioner median: 17973.086
Active median: 23356.895


The mean seems quite higher than the median. This means there could be outliers in the dataset. It's safer to replace the values with the median instead.

<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> Correct

In [45]:
#  Write a function that we will use for filling in missing values
def fill_total_income(value):
    if (value['client_age_group'] == 'Pensioner') and (pd.isna(value['total_income']) == True):
        return pensioner_median
    if (value['client_age_group'] == 'Active') and (pd.isna(value['total_income']) == True):
        return active_median
    else:
        return value['total_income']
    

In [46]:
# Check if it works
row = [
    [1, 'Pensioner', '10', 'school fees'],
    [1, 'Active', '21', 'for fun'],
    [12, 'Active', float('NaN'), 'finally']
]
columns =[ 
'children','client_age_group','total_income','purpose'
]
test = pd.DataFrame(data=row, columns= columns)
test['total_income'] = test.apply(fill_total_income, axis=1)
print(test)



   children client_age_group total_income      purpose
0         1        Pensioner           10  school fees
1         1           Active           21      for fun
2        12           Active    23356.895      finally


In [47]:
# Apply it to every row
df["total_income"] = df.apply(fill_total_income, axis=1)

In [48]:
# Check if we got any errors
print(df['total_income'])

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21403    35966.698
21404    24959.969
21405    14347.610
21406    39054.888
21407    13127.587
Name: total_income, Length: 21408, dtype: float64


It seems to have worked, so let's move on to the next one


In [49]:
# Replacing missing values if there are any errors


Let's check that the total number of values in this column matches the number of values in other ones.

In [50]:
# Checking the number of entries in the columns

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21408 entries, 0 to 21407
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21408 non-null  int64  
 1   days_employed     19308 non-null  float64
 2   dob_years         21408 non-null  int64  
 3   education         21408 non-null  object 
 4   education_id      21408 non-null  int64  
 5   family_status     21408 non-null  object 
 6   family_status_id  21408 non-null  int64  
 7   gender            21408 non-null  object 
 8   income_type       21408 non-null  object 
 9   debt              21408 non-null  int64  
 10  total_income      21408 non-null  float64
 11  purpose           21408 non-null  object 
 12  client_age_group  21408 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB
None


###  Restoring values in `days_employed`

Let's group the data using the previous groups for active workers and pensioners

In [51]:
# Distribution of `days_employed` medians based on your identified parameters
pensioner_days_employed_median = pensioner_category['days_employed'].median()
active_days_employed_median = active_category['days_employed'].median()
print(f'Median days employed for Pensioners: {pensioner_days_employed_median}')
print(f'Median days employed for Active workers: {active_days_employed_median}')


Median days employed for Pensioners: 361032.7835389532
Median days employed for Active workers: 2080.740408224058


In [52]:
# Distribution of `days_employed` means based on your identified parameters
pensioner_days_employed_mean = pensioner_category['days_employed'].mean()
active_days_employed_mean = active_category['days_employed'].mean()
print(f'Mean days employed for Pensioners: {pensioner_days_employed_mean}')
print(f'Mean days employed for Active workers: {active_days_employed_mean}')

Mean days employed for Pensioners: 321223.4365150625
Mean days employed for Active workers: 58168.813000589194


Active workers seem to have more days employed than the pensioners which is not feasible. There must be some outliers in this dataset. It's best we use the median values as they are a better reflection of what it should really be.

In [53]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
def fill_days_employed(value):
    if (value['client_age_group'] == 'Pensioner') and (pd.isna(value['days_employed']) == True):
        return pensioner_days_employed_median
    if (value['client_age_group'] == 'Active') and (pd.isna(value['days_employed']) == True):
        return active_days_employed_median
    else:
        return value['days_employed']
    

In [54]:
# Check that the function works

row = [
    [1, 'Pensioner', '10', 'school fees'],
    [1, 'Active', '21', 'for fun'],
    [12, 'Active', float('NaN'), 'finally']
]
columns =[ 
'children','client_age_group','days_employed','purpose'
]
test = pd.DataFrame(data=row, columns= columns)
test['days_employed'] = test.apply(fill_days_employed, axis=1)
print(test)


   children client_age_group days_employed      purpose
0         1        Pensioner            10  school fees
1         1           Active            21      for fun
2        12           Active   2080.740408      finally


In [55]:
# Apply function to the days_employed
df['days_employed'] = df.apply(fill_days_employed, axis=1)


In [56]:
# Check if function worked

print(df['days_employed'][df['days_employed'].isna()])

Series([], Name: days_employed, dtype: float64)


In [57]:
# Replacing missing values



Let's check that the total number of values in this column matches the number of values in other ones.

In [58]:
# Check the entries in all columns - make sure we fixed all missing values
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21408 entries, 0 to 21407
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21408 non-null  int64  
 1   days_employed     21408 non-null  float64
 2   dob_years         21408 non-null  int64  
 3   education         21408 non-null  object 
 4   education_id      21408 non-null  int64  
 5   family_status     21408 non-null  object 
 6   family_status_id  21408 non-null  int64  
 7   gender            21408 non-null  object 
 8   income_type       21408 non-null  object 
 9   debt              21408 non-null  int64  
 10  total_income      21408 non-null  float64
 11  purpose           21408 non-null  object 
 12  client_age_group  21408 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB
None


<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> Data is ready for analysis. Well done

## Categorization of data
Let's categorize our data based on purpose and number of children. This will help us answer the questions asked and make our conclusions for a scoring system clearer.

In [59]:
# Print the values for your selected data for categorization
print(df['purpose'])

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21403       housing transactions
21404          purchase of a car
21405                   property
21406          buying my own car
21407               to buy a car
Name: purpose, Length: 21408, dtype: object


Let's check unique values

In [60]:
# Check the unique values
print(df['purpose'].unique())
print(df['purpose'][df['purpose'] == 1])
df['purpose'] = df['purpose'].astype('str')

['purchase of the house' 'car purchase' 'supplementary education'
 'to have a wedding' 'housing transactions' 'education' 'having a wedding'
 'purchase of the house for my family' 'buy real estate'
 'buy commercial real estate' 'buy residential real estate'
 'construction of own property' 'property' 'building a property'
 'buying a second-hand car' 'buying my own car'
 'transactions with commercial real estate' 'building a real estate'
 'housing' 'transactions with my real estate' 'cars' 'to become educated'
 'second-hand car purchase' 'getting an education' 'car'
 'wedding ceremony' 'to get a supplementary education'
 'purchase of my own house' 'real estate transactions'
 'getting higher education' 'to own a car' 'purchase of a car'
 'profile education' 'university education'
 'buying property for renting out' 'to buy a car' 'housing renovation'
 'going to university' 1]
291    1
Name: purpose, dtype: object


Based on the unique values, we can see some distinct groups. Wedding related, car related, housing related and education related.

Based on these themes, we will categorize our data.


In [61]:
# Let's write a function to categorize the data based on common topics
def category(value):
    if 'wed' in value:
        return 'wedding related'
    elif 'car' in value:
        return 'car related'
    elif 'real' in value:
        return 'housing related'
    elif 'hous' in value:
        return 'housing related'
    else:
        return 'education related'
    

In [62]:
# Create a column with the categories and count the values for them

df['purpose_group'] = df['purpose'].apply(category)
print(df['purpose_group'].value_counts())

housing related      8252
education related    6539
car related          4295
wedding related      2322
Name: purpose_group, dtype: int64


Let's make categories based on the number of children

In [63]:
# Looking through all the numerical data in your selected column for categorization
print( df['children'])

0        1
1        1
2        0
3        3
4        0
        ..
21403    1
21404    0
21405    1
21406    3
21407    2
Name: children, Length: 21408, dtype: int64


In [64]:
# Getting summary statistics for the column
print(df['children'].unique())

[ 1  0  3  2  4 20  5]


A small family will have at most 2 children, an average family has 4 children at most and anything more than that is a large family.

In [65]:
# Creating function for categorizing into different numerical groups based on ranges

def numeric_category(value):
    if 0 < value <= 2:
        return 'small family'
    elif value == 0:
        return 'no children'
    elif 2 < value <= 4:
        return 'average family'
    else:
        return 'big family'

In [66]:
# Creating column with categories
df['family_type'] = df['children'].apply(numeric_category)

In [67]:
# Count each categories values to see the distribution
print(df['family_type'].value_counts())

no children       14091
small family       6861
average family      371
big family           85
Name: family_type, dtype: int64


<div class="alert alert-block alert-success">
<b>Reviewer's comment v1</b> Good job with categorization

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [68]:
# Check the children data and paying back on time
print(df.pivot_table(index= 'children', values= 'debt', aggfunc='sum'))

# Calculating default-rate based on the number of children
no_children_default = df[(df['family_type'] == 'no children') & (df['debt'] == 1)]
no_children_default = no_children_default['family_type'].count() / df['family_type'].count()

small_family_default = df[(df['family_type'] == 'small family') & (df['debt'] == 1)]
small_family_default = small_family_default['family_type'].count() / df['family_type'].count()

average_family_default = df[(df['family_type'] == 'average family') & (df['debt'] == 1)]
average_family_default = average_family_default['family_type'].count() / df['family_type'].count()

big_family_default = df[(df['family_type'] == 'big family') & (df['debt'] == 1)]
big_family_default = big_family_default['family_type'].count() / df['family_type'].count()

print(f'Percentage of families with no children defaults : {no_children_default:.3%}')

print(f'Percentage of small family defaults : {small_family_default:.3%}')

print(f'Percentage of average family defaults : {average_family_default:.3%}')

print(f'Percentage of big family defaults : {big_family_default:.3%}')




          debt
children      
0         1063
1          445
2          194
3           27
4            4
5            0
20           8
Percentage of families with no children defaults : 4.965%
Percentage of small family defaults : 2.985%
Percentage of average family defaults : 0.145%
Percentage of big family defaults : 0.037%


**Conclusion**

From our analysis we can see that the customers with less children have the highest default rates. So the more the children, the less the chances of them defaulting on their loans.


<div class="alert alert-block alert-warning">
<b>Reviewer's comment v1</b> Great that you work with pivot tables! If we add aggregation by mean, we can simply find share:
    
        df.pivot_table(index= 'children', values= 'debt', aggfunc=['count', 'sum', 'mean'])
        

**Is there a correlation between family status and paying back on time?**

In [69]:
# Check the family status data and paying back on time

print(df.pivot_table(index= 'family_status', values= 'debt', aggfunc='sum'))


# Calculating default-rate based on family status

civil_partnership_default = df[(df['family_status'] == 'civil partnership') & (df['debt'] == 1)]
civil_partnership_default = civil_partnership_default['family_status'].count() / df['family_status'].count()

divorced_default = df[(df['family_status'] == 'divorced') & (df['debt'] == 1)]
divorced_default = divorced_default['family_status'].count() / df['family_status'].count()

married_default = df[(df['family_status'] == 'married') & (df['debt'] == 1)]
married_default = married_default['family_status'].count() / df['family_status'].count()

unmarried_default = df[(df['family_status'] == 'unmarried') & (df['debt'] == 1)]
unmarried_default = unmarried_default['family_status'].count() / df['family_status'].count()

widow_widower_default = df[(df['family_status'] == 'widow / widower') & (df['debt'] == 1)]
widow_widower_default = widow_widower_default['family_status'].count() / df['family_status'].count()

print(f'Percentage of civil partnerships defaults : {civil_partnership_default:.3%}')

print(f'Percentage of divorced defaults : {divorced_default:.3%}')

print(f'Percentage of married defaults : {married_default:.3%}')

print(f'Percentage of unmarried defaults : {unmarried_default:.3%}')

print(f'Percentage of widow / widower defaults : {widow_widower_default:.3%}')



                   debt
family_status          
civil partnership   389
divorced             85
married             930
unmarried           274
widow / widower      63
Percentage of civil partnerships defaults : 1.817%
Percentage of divorced defaults : 0.397%
Percentage of married defaults : 4.344%
Percentage of unmarried defaults : 1.280%
Percentage of widow / widower defaults : 0.294%


**Conclusion**

Based on the results, we can see that married customers are more likely to default on a loan than any other group of people.

**Is there a correlation between income level and paying back on time?**

In [70]:
# Check the income level data and paying back on time

print(df.pivot_table(index= 'income_type', values= 'debt', aggfunc='sum'))


# Calculating default-rate based on income level

business_income_default = df[(df['income_type'] == 'business') & (df['debt'] == 1)]
business_income_default = business_income_default['income_type'].count() / df['income_type'].count()

civil_servant_income_default = df[(df['income_type'] == 'civil servant') & (df['debt'] == 1)]
civil_servant_income_default = civil_servant_income_default['income_type'].count() / df['income_type'].count()

employee_income_default = df[(df['income_type'] == 'employee') & (df['debt'] == 1)]
employee_income_default = employee_income_default['income_type'].count() / df['income_type'].count()

entrepreneur_income_default = df[(df['income_type'] == 'entrepreneur') & (df['debt'] == 1)]
entrepreneur_income_default = entrepreneur_income_default['income_type'].count() / df['income_type'].count()

leave_income_default = df[(df['income_type'] == 'paternity / maternity leave') & (df['debt'] == 1)]
leave_income_default = leave_income_default['income_type'].count() / df['income_type'].count()

retiree_income_default = df[(df['income_type'] == 'retiree') & (df['debt'] == 1)]
retiree_income_default = retiree_income_default['income_type'].count() / df['income_type'].count()

student_income_default = df[(df['income_type'] == 'student') & (df['debt'] == 1)]
student_income_default = student_income_default['income_type'].count() / df['income_type'].count()

unemployed_income_default = df[(df['income_type'] == 'unemployed') & (df['debt'] == 1)]
unemployed_income_default = unemployed_income_default['income_type'].count() / df['income_type'].count()

unknown_income_default = df[(df['income_type'] == 'unknown') & (df['debt'] == 1)]
unknown_income_default = unknown_income_default['income_type'].count() / df['income_type'].count()


print(f'Percentage of business income defaults : {business_income_default:.3%}')

print(f'Percentage of civil servant income defaults : {civil_servant_income_default:.3%}')

print(f'Percentage of employee income defaults : {employee_income_default:.3%}')

print(f'Percentage of entrepreneur income defaults : {entrepreneur_income_default:.3%}')

print(f'Percentage of leave income defaults : {leave_income_default:.3%}')

print(f'Percentage of retiree income defaults : {retiree_income_default:.3%}')

print(f'Percentage of student income defaults : {student_income_default:.3%}')

print(f'Percentage of unemployed income defaults : {unemployed_income_default:.3%}')

print(f'Percentage of unknown income defaults : {unknown_income_default:.3%}')

                             debt
income_type                      
business                      376
civil servant                  86
employee                     1060
entrepreneur                    0
paternity / maternity leave     1
retiree                       216
student                         0
unemployed                      1
unknown                         1
Percentage of business income defaults : 1.756%
Percentage of civil servant income defaults : 0.402%
Percentage of employee income defaults : 4.951%
Percentage of entrepreneur income defaults : 0.000%
Percentage of leave income defaults : 0.005%
Percentage of retiree income defaults : 1.009%
Percentage of student income defaults : 0.000%
Percentage of unemployed income defaults : 0.005%
Percentage of unknown income defaults : 0.005%


**Conclusion**

From the above data, we can conclued that employees are far more likely to default on a loan than any other group of individuals.

**How does credit purpose affect the default rate?**

In [71]:
# Check the percentages for default rate for each credit purpose and analyze them
car_related_default = df[(df['purpose_group'] == 'car related') & (df['debt'] == 1)]
car_related_default = car_related_default['purpose_group'].count() / df['purpose_group'].count()

education_related_default = df[(df['purpose_group'] == 'education related') & (df['debt'] == 1)]
education_related_default = education_related_default['purpose_group'].count() / df['purpose_group'].count()

housing_related_default = df[(df['purpose_group'] == 'housing related') & (df['debt'] == 1)]
housing_related_default = housing_related_default['purpose_group'].count() / df['purpose_group'].count()

wedding_related_default = df[(df['purpose_group'] == 'wedding related') & (df['debt'] == 1)]
wedding_related_default = wedding_related_default['purpose_group'].count() / df['purpose_group'].count()

print(df.pivot_table(index= 'purpose_group', values= 'debt', aggfunc='sum'))


print(f'Percentage of car related defaults : {car_related_default:.3%}')

print(f'Percentage of education related defaults : {education_related_default:.3%}')

print(f'Percentage of housing related defaults : {housing_related_default:.3%}')

print(f'Percentage of wedding related defaults : {wedding_related_default:.3%}')




                   debt
purpose_group          
car related         402
education related   561
housing related     592
wedding related     186
Percentage of car related defaults : 1.878%
Percentage of education related defaults : 2.621%
Percentage of housing related defaults : 2.765%
Percentage of wedding related defaults : 0.869%


**Conclusion**

From our analysis, we can conclude that customers who take loans for wedding related purposes are more likely to pay back.

<div class="alert alert-success">
<b>Reviewer's comment v1</b> Well done

# General Conclusion 
In summary, the dataset contained information about different customers seeking loans. 10% of the data in `days_employed` and `total_income` columns were missing. As a matter of fact, the rows with missing values were the same in both columns which led us to believe that they could be from people who were unemployed. Also 70% of the data in `days_employed` column was negative and so had to be changed to positive as there is no such thing as negative days employed.

As for the missing data, the data was split into 2 groups, active workers and pensioners using the national retirment age of 65 as the determinant for this split. The rows were then filled with the median value of the columns which they fell under. The median was chosen as the value to fill because these rows had outliers evidenced by their non-coherent means.

From our analysis, the smaller the family, the more likelihood that they would default on their loans. This was the most consistent conclusion. As for other parameters, higher default rates were associated with married individuals and employees. On the other hand, customers who took loans for wedding related reasons were shown to have a higher tendency to pay back.

And so with this, we can make our credit score system using a scale of 5 for each correlation. The higher the repayment rate in each group, the more the points.



<div class="alert alert-success">
<b>Reviewer's comment v1:</b>
    
Thanks for the final conclusion! You conducted great research.